### ISSUE TITLE
Integration with Spark

### ISSUE THREAD

**Issue by daniel-acuna** [glm-tools/pyglmnet#210]
Hey @pavanramkumar and @jasmainak, great work here! 

I would like to extend this package to work with [Apache Spark](spark.apache.org). I am wondering if there is an easy way of combining two [GLM](https://github.com/glm-tools/pyglmnet/blob/master/pyglmnet/pyglmnet.py#L222) objects, one trained in one part of the data and another in another part of the data. That would make a map-reduce job very easy. Any thoughts?

Thanks!

**Comment by pavanramkumar** [comment]
Hey @daniel-acuna awesome initiative! 

The `GLM()` object now fits the model for a single value of `reg_lambda` and stores the fit in (`self.beta0_`, `self.beta_`).

I believe if you wanted to implement the model to streaming data, you would warm-initialize the parameters for the new mini-batch with the parameters of the old mini-batch. This should be feasible and I would recommend doing it in a new class: see how we do it for warm restarts in [`GLMCV()`](https://github.com/glm-tools/pyglmnet/blob/master/pyglmnet/pyglmnet.py#L964-L1011).

P. S. Note that we removed `multinomial` from the offerings in the `GLM()` class since the shape issues were adding significant developer overhead.

**Comment by daniel-acuna** [comment]
I see that the API is much improved now!  I guess map-reduce would go beyond warm restarts. Any objection to just taking the average of `beta0_` and `beta_` from two `GLM` objects to merge them?
P.S. Sad to see `multinomial` go but I can see how it makes things more complicated. Spark has multinomial regression with ElasticNet regularization but I am really interested in the Tikhonov specification. Really cool

**Comment by pavanramkumar** [comment]
Oh because you'd fit the batches in parallel? I would try to follow the exact same conventions used in [mllib](https://spark.apache.org/docs/latest/mllib-linear-methods.html) and try not to change anything.

Regarding averaging `beta0_` and `beta_` from multiple parallel batches, I guess it is ok for a convex problem like GLM, but you see how this would be bad for non-convex problems. I'd recommend writing a test to benchmark that you get the same solution on a small simulated dataset using parallel methods.

I'd also take a look at the [lightning](http://contrib.scikit-learn.org/lightning/intro.html) package to see if you can borrow ideas for optimization from SAG, SAGA, SVRG...

P. S. Yes, more than the link functions, Tikhonov and group lasso have become the USPs of this package. Also, we have agreed to bring back `multinomial` with a new class `MultinomialGLM()` and `MultinomialGLMCV()`. Similar strategy for other link functions like `quasi-poisson` that have more parameters.

**Comment by jasmainak** [comment]
@pavanramkumar @daniel-acuna would this issue be a duplicate of https://github.com/glm-tools/pyglmnet/issues/136 ?

**Comment by daniel-acuna** [comment]
#136 is about building on top of Thunder, which depends on Spark. #210 is more about building something  directly related to Spark. I would prefer the latter?

**Comment by pavanramkumar** [comment]
Yes, I think #136 is low priority: Thunder is not as actively developed as it used to be.

**Comment by jasmainak** [comment]
I'm closing this issue as it has not happened in 2 years ...
